src/mlpack/methods/cf/interpolation_policies/regression_interpolation.hpp

```diff
@@ -0,0 +1,171 @@
+   * constant terms for the equations and then solve the equations. The
+   * solution of the linear system of equations is the resulting interpolation
+   * weights (the first parameter). After getting the weights, CF algorithm
+   * multiplies each neighbor's rating by its corresponding weight and sums
+   * them to get predicted rating.
+   *
+   * @param weights Resulting interpolation weights. The size of weights should
+   *     be set to the number of neighbors before calling GetWeights().
+   * @param w Matrix W from decomposition.
+   * @param h Matrix H from decomposition.
+   * @param queryUser Queried user.
+   * @param neighbors Neighbors of queried user.
+   * @param similarities Similarites between query user and neighbors.
+   * @param cleanedData Sparse rating matrix.
+   */
+  template <typename VectorType>
+  void GetWeights(VectorType&& weights,
+                  const arma::mat& w,
+                  const arma::mat& h,
+                  const size_t queryUser,
+                  const arma::Col<size_t>& neighbors,
+                  const arma::vec& /* similarities*/,
+                  const arma::sp_mat& cleanedData)
+  {
+    if (weights.n_elem != neighbors.n_elem)
+    {
+      Log::Fatal << "The size of the first parameter (weights) should "
+          << "be set to the number of neighbors before calling GetWeights()."
+          << std::endl;
+    }
+
+    const size_t itemNum = cleanedData.n_rows;
+    const size_t neighborNum = neighbors.size();
+
+    // Coeffcients of the linear equations used to compute weights.
+    arma::mat coeff(neighborNum, neighborNum);
+    // Constant terms of the linear equations used to compute weights.
+    arma::vec constant(neighborNum);
+
+    arma::vec userRating(cleanedData.col(queryUser));
+    const size_t support = arma::accu(userRating != 0);
+
+    // If user has no rating at all, average interpolation is used.
+    if (support == 0)
+    {
+      weights.fill(1.0 / neighbors.n_elem);
+      return;
+    }
+
+    for (size_t i = 0; i < neighborNum; i++)
+    {
+      // Calculate coefficient.
+      arma::vec iPrediction;
+      for (size_t j = i; j < neighborNum; j++)
+      {
+        if (a(neighbors(i), neighbors(j)) != 0)
+        {
+          // The coefficient has already been cached.
+          coeff(i, j) = a(neighbors(i), neighbors(j));
+          coeff(j, i) = coeff(i, j);
+        }
+        else
+        {
+          // Calculate the coefficient.
+          if (iPrediction.size() == 0)
+            // Avoid recalculation of iPrediction.
+            iPrediction = w * h.col(neighbors(i));
+          arma::vec jPrediction = w * h.col(neighbors(j));
+          coeff(i, j) = arma::dot(iPrediction, jPrediction) / itemNum;
+          if (coeff(i, j) == 0)
+            coeff(i, j) = std::numeric_limits<double>::min();
+          coeff(j, i) = coeff(i, j);
+          // Cache calcualted coefficient.
+          a(neighbors(i), neighbors(j)) = coeff(i, j);
```

Comment:
I am not sure that it is reasonable to cache the coefficients since armadillo stores sparse matrices in CSC format, and populating a sparse matrix via element access operators is inefficient. Moreover you don't take into account that the matrix is symmetric. I am not sure but I think `std::map` could handle that better. How do you think?

Reply:
Ah I understand. Thanks for pointing this out. I didn't know how sparse matrix is implemented and I will take a look at CSC format later. I think `std::unordered_map` may be a better choice because it is generally faster than `std::map`.

Reply:
Actually, this has changed recently (for Armadillo 7+ I believe).  We implemented a 'cache' for Armadillo so that setting sparse matrix elements by, e.g., `x(a, b) = c` is faster.  Basically in the backend this fills out a `std::map<>`, which is synced with the actual CSC representation when necessary.  I don't have any particular opinions on how you do this here, but I thought I'd let you know since I saw the comment.

If you're interested in more details, here is a paper on how it works: http://www.ratml.org/pub/pdf/2018open.pdf

Reply:
@rcurtin Thanks for the clear explanation:)
@lozhnikov In this case I think we can keep using sparse matrix to cache the values, right?

Reply:
@rcurtin Thanks for the clarification, it's interesting.
Actually, looks like the correct link is http://www.ratml.org/pub/pdf/2018user.pdf :)
@Wenhao-H Yeah, sure.

Reply:
Oops, right, looked like I typed it wrong.  Thanks for the fix :)